v2/pkg/image/image.go

```diff
@@ -5,35 +5,94 @@ import (
 	"strings"
 )
 
-func IsImageByDigest(imgRef string) bool {
-	return strings.Contains(imgRef, "@")
+// specification is sourced from github.com/containers/image/blob/main/docker/reference/reference.go
+// Grammar
+//
+//	reference                       := name [ ":" tag ] [ "@" digest ]
+//	name                            := [domain '/'] path-component ['/' path-component]*
+//	domain                          := domain-component ['.' domain-component]* [':' port-number]
+//	domain-component                := /([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/
+//	port-number                     := /[0-9]+/
+//	path-component                  := alphanumeric [separator alphanumeric]*
+//	alphanumeric                   := /[a-z0-9]+/
+//	separator                       := /[_.]|__|[-]*/
+//
+//	tag                             := /[\w][\w.-]{0,127}/
+//
+//	digest                          := digest-algorithm ":" digest-hex
+//	digest-algorithm                := digest-algorithm-component [ digest-algorithm-separator digest-algorithm-component ]*
+//	digest-algorithm-separator      := /[+.-_]/
+//	digest-algorithm-component      := /[A-Za-z][A-Za-z0-9]*/
+//	digest-hex                      := /[0-9a-fA-F]{32,}/ ; At least 128 bit digest value
+//
+//	identifier                      := /[a-f0-9]{64}/
+//	short-identifier                := /[a-f0-9]{6,64}/
+type ImageSpec struct {
+	Transport              string
+	Reference              string
+	ReferenceWithTransport string
+	Name                   string
+	Domain                 string
+	PathComponent          string
+	Tag                    string
+	Digest                 string
 }
 
-func PathWithoutDNS(imgRef string) (string, error) {
-
-	var imageName []string
-	if IsImageByDigest(imgRef) {
-		imageNameSplit := strings.Split(imgRef, "@")
-		imageName = strings.Split(imageNameSplit[0], "/")
+// It expects the image reference not to have the transport prefix.
+// Otherwise, it will return an error.
+func ParseRef(imgRef string) (ImageSpec, error) {
+	var imgSpec ImageSpec
+	if strings.Contains(imgRef, "://") {
+		imgSpec.ReferenceWithTransport = imgRef
+		imgSplit := strings.Split(imgRef, "://")
+		imgSpec.Transport = imgSplit[0] + "://"
```

Comment:
Should we do a len check here - just to avoid ugly panics ?

Reply:
This is fenced behind the `strings.Contains`. Do you think we should add another if ?